app/lib/common/utilities/genome_data.dart

```diff
@@ -0,0 +1,77 @@
+import 'dart:convert';
+
+import 'package:http/http.dart';
+
+import '../../profile/models/hive/alleles.dart';
+import '../../profile/models/hive/cpic_lookup_response.dart';
+import '../../profile/models/hive/diplotype.dart';
+import '../constants.dart';
+import '../services.dart';
+
+Future<void> fetchAndSaveAllesData(String token, String url) async {
+  final userAlleleData = getBox<Alleles>(Boxes.alleles);
```

Comment:
I've been doing some thinking on how we can best address this. I agree that it's kind of annoying to have to deal with so many boxes all the time, it makes it hard to keep track. I am also not a huge fan of how we currently have to interface with the boxes, to me it is a little verbose.

I also, however, am not a huge fan of just putting alleles, lookups, and lastFetch in a single box as this would result in a loss of strong typing and we would have to cast objects everytime we pull something from the box.

As a solution, I was thinking that perhaps creating some sort of global `userData` structure (class) which can house everything pertaining to the user might be a possibility (i.e, everything which is individual to each person using the app). This includes: star allele data from the lab server, relevant lookups keys, metadata on last login / fetch with the possibility of more down the line.

Using this we could reduce the number of boxes that we use down to only one box which houses a single instance of type `userData`. The idea would be that we would open this box exactly once at runtime and export the object as some sort of "global" userData / userConfig store which other services can interface with.

For instance, this might result in something like this in `services.dart` (where we initialize hive + boxes).

```dart
// get the first (and only) instance of userData from hive
final userData = Hive.openBox<userData>("userConfig").getAt(0);
//...
```

which may be called in other services by:

```dart
import '../some/path/to/services.dart';

final alleles = userData.genomicData.diplotypes;
final lookups = userData.genomicData.annotationLookups;
//...
final isLoggedIn = userData.metaData.isLoggedIn;

```

I personally think this may help in managing the box complexity as we add new bits and pieces of data which we need to save down the line, as currently you either create a new box for every type or you end up having to cast objects all throughout the codebase.

It's also annoying having to remember all the box names (although we did somewhat address this using enums for the box names, however the problem still remains when inserting / fetching specific keys). This issue would also be solved as we would not have to ever directly interact with the boxes (except for once in `services.dart`). Furthermore we would also retain strong typing.

It also reads more nicely IMO that having to see something like:
```dart
await Hive.openBox<UserAlleles>(Boxes.userAlleles.toString()).putAt(0, alleles);
```
every couple of lines 🤣. Instead, I imagine something more along the lines of:
```dart
await userData.genomicData.updateAlleles(alleles);
```

Do let me know what you think of this proposal, although I am open to suggestions.

Also, sorry in advance for forcing you to read this novel of a response, just wanted to articulate it as best as I could.

Reply:
I can also agree with @Toalaah, the current reason to have so many boxes is the typing. Should we put all of them into one, then the return type is just `dynamic` :(

Reply:
Resolving this conversation as this subject will be tackled in a seperate branch  / MR. See the following issue here: #225

Reply:
I do agree we need to reduce boxes, but I think we should think semantically about what type of box makes sense. So I wouldn't blend everything into 1 box but start with all the boxes we have and try to put them together (by making custom aggregate data structures and making boxes that store those)

Everything that is about to come up is only valid under the assumption that I understood our proposed information flow (#199). Please correct me if any of this is wrong @jannis-baum.

First of all I would refrain entirely from using the term 'allele' in our code. We are never working with alleles, we are sent diplotypes straight away, which mostly already blend many alleles together I think.

Now for a 'gene symbol' (which I am pretty sure is mostly not a single gene but a blend of many genes again), there are a bunch of diplotypes, a bunch of variations that this blueprint of a 'genesymbol' can have. But big subsets of these diploytpes within a single 'genesymbol' are pharmacogenomically equivalent. So we don't need to store the diplotype necessarily but just the pharmacogenomically relevant part (this is the mapping from diplotype(s) to 'lookupkeys')

So a 'genesymbol' like CYP2C19 may have 'lookupkeys' like Normal/Intermediate Metabolizer. This is the information that we need to effectively store.

So my first proposition for a custom data structure would look something like this:

`class userLookup {
  string genesymbol;
  string lookupkey;
}`

Since a human can only have one diplotype for each 'genesymbol', and there is a n:1 mapping from diplotypes to lookupkeys, a human can only have one 'lookupkey' for each 'genesymbol'. If that is wrong, we just use `List<string> lookupkeys;` instead.
I think a box of this data type makes the most sense semantically. As for the other (previous) boxes:

- userData (formerly alleles): I am pretty sure this is a remnant of some sort. If I understand the proposed information flow (#199) right, we don't (need to) save data this specific at all. We do work with data of this sort but IMO we should parse it straight to these genesymbol-lookupkey-pairs.
- lookups: The contents of this box would pretty much be covered by the userLookup box. The difference here is that the box actually contains a bunch of lookups instead of just one List of all lookups...
- lastLookupSomething: The way this is implemented now, it stores the last DateTime ALL lookups were fetched. If it remains that way I would integrate it in the preferences box (see below), if not, I would integrate it into the userLookup data structure `DateTime lastFetched;`
- preferences: This box is a bit more interesting because it should be able to store all kinds of information. I would like to avoid no typing and casting at any cost, my first idea would be to make a custom data object 'Preference' that contains everything that we need, for example:
    ```
    class Preference {
      DateTime lastLookupsFetch;
      bool isLoggedIn;
    }
    ```
    Then we could just store one Preference object in the preferences box. But whenever we just store one object in a box, we should be thinking: 'Why use a box at all?' We could solve this with a singleton or something, probably there is even a flutter native feature, that supports saving preference-like data.
    
    I do like the idea to rename 'preferences' to something else like 'metaData' though. isLoggedIn is not a preference IMO and lastLookupsFetch wouldn't be one either.

I would also like to apologize for this long-ass answer :D